hcl2template/parser.go

```diff
@@ -293,112 +272,311 @@ func filterVarsFromLogs(inputOrLocal Variables) {
 	}
 }
 
-func (cfg *PackerConfig) Initialize(opts packer.InitializeOptions) hcl.Diagnostics {
-	diags := cfg.InputVariables.ValidateValues()
-	diags = append(diags, cfg.LocalVariables.ValidateValues()...)
-	diags = append(diags, cfg.evaluateDatasources(opts.SkipDatasourcesExecution)...)
-	diags = append(diags, checkForDuplicateLocalDefinition(cfg.LocalBlocks)...)
-	diags = append(diags, cfg.evaluateLocalVariables(cfg.LocalBlocks)...)
+// decodeFile attempts to decode the configuration from a HCL file, and starts
+// populating the config from it.
+func (cfg *PackerConfig) decodeFile(file *hcl.File) hcl.Diagnostics {
+	var diags hcl.Diagnostics
 
-	filterVarsFromLogs(cfg.InputVariables)
-	filterVarsFromLogs(cfg.LocalVariables)
+	content, moreDiags := file.Body.Content(configSchema)
+	diags = append(diags, moreDiags...)
+	// If basic parsing failed, we should not continue
+	if diags.HasErrors() {
+		return diags
+	}
 
-	// parse the actual content // rest
-	for _, file := range cfg.files {
-		diags = append(diags, cfg.parser.parseConfig(file, cfg)...)
+	for _, block := range content.Blocks {
+		diags = append(diags, cfg.decodeBlock(block)...)
 	}
 
-	diags = append(diags, cfg.initializeBlocks()...)
+	return diags
+}
+
+func (cfg *PackerConfig) decodeBlock(block *hcl.Block) hcl.Diagnostics {
+	switch block.Type {
+	case packerLabel:
+		return cfg.decodePackerBlock(block)
+	case dataSourceLabel:
+		return cfg.decodeDatasource(block)
+	case variableLabel:
+		return cfg.decodeVariableBlock(block)
+	case variablesLabel:
+		return cfg.decodeVariablesBlock(block)
+	case localLabel:
+		return cfg.decodeLocalBlock(block)
+	case localsLabel:
+		return cfg.decodeLocalsBlock(block)
+	// NOTE: Both build and source blocks can be dynamically expanded.
+	//
+	// This means that while at this time we can already get some information
+	// about them, we may not be able to decode their final form at this time
+	// and we can only do so when their dependencies are evaluated.
```

Comment:
This is my suggested comment based off what I take your comment to mean. 

That said is this comment not true for any block that can contain dynamic data such as locals or variables? 
```suggestion
	// NOTE: Both build and source blocks can be dynamically expanded.
	//
	// This means that only the known configuration options that do not include 
	// expressions or dynamic blocks information will be decoded. Unknown dependencies
	// will be decoded in a second pass by calling (finalizeDecode or finalizeDecodeSource) only after all resources has been evaluated.
```

Reply:
Locals and variables (and datasources) cannot contain `dynamic` blocks as of now. For variables and locals (both singular or plural), that wouldn't make sense to support it even, as the structure is completely static. For datasources, it is just a matter of when we expand the blocks, and right now, it happens during `Initialize`, after we've evaluated the datasources and local variables, so while it would make sense eventually to support them, it is not yet a valid use-case for Packer.